src/session_server.c

```diff
@@ -1400,6 +1400,186 @@ recv_rpc_check_msgid(struct nc_session *session, const struct lyd_node *envp)
+    }
+    cn = lys_find_path(ctx, NULL, str, 0);
+    free(str);
+
+    return cn;
+}
+
+/**
+ * @brief Find the nth substring delimited by quotes.
+ *
+ * For example: abcd"ef"ghij"kl"mn -> rank 1 is "ef", rank 2 is "kl".
+ *
+ * @param[in] msg Input string with quoted substring.
+ * @param[in] rank Number starting from 1 specifying the nth substring.
+ * @return Copied nth substring without quotes.
+ */
+static char *
+nc_rpc_err_get_quoted_string(const char *msg, uint32_t rank)
+{
+    char *ret;
+    const char *start = NULL, *end = NULL, *iter, *tmp;
+    uint32_t cnt = 0;
+
+    assert(msg && (rank > 0));
+
+    rank *= 2;
+    for (iter = msg; *iter != '\0'; ++iter) {
+        if (*iter == '\"') {
+            /* updating the start and end pointers */
+            tmp = end;
+            end = iter;
+            start = tmp;
+            if (++cnt == rank) {
+                /* nth substring found */
+                break;
+            }
+        }
+    }
+
+    if (!start) {
+        return NULL;
+    }
+
+    /* Skip first quote */
+    ++start;
+    /* Copy substring */
+    ret = strndup(start, end - start);
+
+    return ret;
+}
+
+/**
+ * @brief Check that the @p str starts with the @p prefix.
+ *
+ * @param[in] prefix Required prefix.
+ * @param[in] str Input string to check.
+ * @return True if @p str start with @p prefix otherwise False.
+ */
+static ly_bool
+nc_strstarts(const char *prefix, const char *str)
+{
+    return strncmp(str, prefix, strlen(prefix)) == 0;
+}
+
+/**
+ * @brief Prepare reply for rpc error.
+ *
+ * @param[in] session NETCONF session.
+ * @param[in] envp NETCONF-specific RPC envelope. Can be NULL.
+ * @return rpc-reply object or NULL.
+ */
+static struct nc_server_reply *
+nc_server_prepare_rpc_err(struct nc_session *session, struct lyd_node *envp)
+{
+    struct lyd_node *reply = NULL;
+    const struct lysc_node *cn;
+    const struct ly_err_item *ly_err;
+    NC_ERR_TYPE errtype;
+    const char *attr;
+    char *str = NULL, *errmsg = NULL, *schema_path = NULL;
+    LY_ERR errcode;
+
+    /* envelope was not parsed */
+    if (!envp && (session->version != NC_VERSION_11)) {
+        return NULL;
+    }
+    ly_err = ly_err_last(session->ctx);
+    if (!envp && !strcmp("Missing XML namespace.", ly_err->msg)) {
+        reply = nc_err(session->ctx, NC_ERR_MISSING_ATTR, NC_ERR_TYPE_RPC, "xmlns", "rpc");
+        goto cleanup;
+    } else if (!envp) {
+        /* completely malformed message, NETCONF version 1.1 defines sending error reply from
+         * the server (RFC 6241 sec. 3) */
+        reply = nc_err(session->ctx, NC_ERR_MALFORMED_MSG);
+        return nc_server_reply_err(reply);
+    }
+    /* at least the envelopes were parsed */
+    assert(envp);
+
+    /* store strings, to avoid overwriting ly_err */
+    errmsg = strdup(ly_err->msg);
+    if (!errmsg) {
+        reply = nc_err(session->ctx, NC_ERR_OP_FAILED, NC_ERR_TYPE_APP);
+        goto cleanup;
+    }
+    assert(ly_err->msg);
```

Comment:
Seems redundant at this point, after `strdup()` is called on it, it crashes on `NULL` I believe.